src/apps/peer/udpserver.c

```diff
@@ -32,86 +32,123 @@
 #include "apputils.h"
 #include "stun_buffer.h"
 
+struct listen_info {
+  ioa_addr addr;
+  int received;
+  int verbose;
+  struct listen_info *next;
+};
+
 /////////////// io handlers ///////////////////
 
 static void udp_server_input_handler(evutil_socket_t fd, short what, void *arg) {
 
   if (!(what & EV_READ))
     return;
 
-  ioa_addr *addr = (ioa_addr *)arg;
+  struct listen_info *listen = (struct listen_info *)arg;
 
   int len = 0;
-  int slen = get_ioa_addr_len(addr);
-  stun_buffer buffer;
-  ioa_addr remote_addr;
+  int slen = get_ioa_addr_len(&listen->addr);
+  stun_buffer *buffer = (stun_buffer *)malloc(sizeof(stun_buffer));
+  ioa_addr remote_addr = {0};
+  if (!buffer)
+    return;
 
   do {
-    len = recvfrom(fd, buffer.buf, sizeof(buffer.buf) - 1, 0, (struct sockaddr *)&remote_addr, (socklen_t *)&slen);
+    len = recvfrom(fd, buffer->buf, sizeof(buffer->buf) - 1, 0, (struct sockaddr *)&remote_addr, (socklen_t *)&slen);
   } while (len < 0 && socket_eintr());
```

Comment:
This is attempting to read until `recvfrom` stops failing? recvfrom returns a positive number representing how many bytes were read, or a negative number on failure, and 0 on "no messages available".

I find the way this is setup to be really weird, but maybe that's just the design and i'm not familiar enough with the project to understand yet.